pkg/daemon/utils/ip_range.go

```diff
@@ -0,0 +1,226 @@
+			// add [cidrStart, currentRangeStartPrev] to exclude ip ranges
+			currentRangeStartPrev := ip.PrevIP(currentIPRange.start)
+
+			ipRange, err := CreateIPRange(cidrStart, currentRangeStartPrev)
+			if err != nil {
+				return nil, fmt.Errorf("create ip range for %v~%v failed: %v", cidrStart, currentRangeStartPrev, err)
+			}
+
+			if ipRange != nil {
+				excludeIPRanges = append(excludeIPRanges, ipRange)
+			}
+		}
+
+		var nextRangeStartPrev net.IP
+		if currentIPRangeIndex == len(includedRanges)-1 {
+			nextRangeStartPrev = cidrEnd
+		} else {
+			nextRangeStartPrev = ip.PrevIP(includedRanges[currentIPRangeIndex+1].start)
+		}
+
+		// add [endNext, nextRangeStartPrev] to exclude ip ranges
+		endNext := ip.NextIP(currentIPRange.end)
+
+		ipRange, err := CreateIPRange(endNext, nextRangeStartPrev)
+		if err != nil {
+			return nil, fmt.Errorf("create ip range for %v~%v failed: %v", endNext, nextRangeStartPrev, err)
+		}
+
+		if ipRange != nil {
+			excludeIPRanges = append(excludeIPRanges, ipRange)
+		}
+	}
+
+	allExcludedIPs := excludeIPs
+	if len(reservedIPs) != 0 {
+		allExcludedIPs = append(allExcludedIPs, reservedIPs...)
+	}
+
+	if gateway != nil {
+		allExcludedIPs = append(allExcludedIPs, gateway)
+	}
+
+Loop2:
+	for _, ipAddr := range allExcludedIPs {
+		if len(excludeIPRanges) != 0 {
+			for _, ipRange := range excludeIPRanges {
+				if ipRange.TryAddIP(ipAddr) {
+					continue Loop2
+				}
+			}
+		}
+
+		singleIPRange, err := CreateIPRange(ipAddr, ipAddr)
+		if err != nil {
+			return nil, fmt.Errorf("create ip range for ip %v failed: %v", ipAddr, err)
+		}
+		excludeIPRanges = append(excludeIPRanges, singleIPRange)
+	}
+
+	var excludeIPBlocks []*net.IPNet
+	for _, ipRange := range excludeIPRanges {
+		excludeIPBlocks = append(excludeIPBlocks, ipRange.splitIPRangeToIPBlocks()...)
+	}
+
+	return excludeIPBlocks, nil
+}
+
+func (ir *IPRange) splitIPRangeToIPBlocks() []*net.IPNet {
+	rangeStart := ir.start
+	rangeEnd := ir.end
+	nextRangeStart := rangeStart
+
+	var ipBlocks []*net.IPNet
+
+	for nextRangeStart != nil {
+		var newBlock *net.IPNet
+		newBlock, nextRangeStart = findTheFirstLargestCidr(nextRangeStart, rangeEnd)
+		ipBlocks = append(ipBlocks, newBlock)
+	}
+
+	return ipBlocks
+}
+
+func calculateIPLastZeroBits(ip net.IP) int {
+	zeroBits := 0
+	for {
+		nextPossibleZeroBits := zeroBits + 1
+		ipInt := ipToInt(ip)
+		ipUint64 := ipInt.Uint64()
+
+		if ipUint64 == (ipUint64>>nextPossibleZeroBits)<<nextPossibleZeroBits {
+			zeroBits = nextPossibleZeroBits
+		} else {
+			break
+		}
+	}
+
+	return zeroBits
+}
+
+func findTheFirstLargestCidr(start, end net.IP) (*net.IPNet, net.IP) {
+	// The max possible cidr size for the start ip to represent.
+	zeroBits := calculateIPLastZeroBits(start)
+	var ipLen = net.IPv4len * 8
+	if start.To4() == nil {
+		ipLen = net.IPv6len * 8
+	}
+
+	minCidrPrefixLen := ipLen - zeroBits
+	maxValidCidrPrefixLen := minCidrPrefixLen
+
+	for {
+		tmpCidr := &net.IPNet{
+			IP:   start,
```

Comment:
should IP be masked by Mask to be a standard IPNet?

Reply:
No need to do this, the last `zeroBits` must be zero, we don't need to mask it again.

Reply:
Mask is changing every loop, so we could get some nonstandard CIDR like 192.168.0.4/24, 192.168.0.4/25, 192.168.0.4/26 ...

Reply:
mistake, gotcha your point